File: Kakao/2018_방금그곡.cpp
// O(musicinfo.size * lengthOfContent)

#include <string>
#include <vector>
#include <iostream>

using namespace std;

struct MusicInfo
{
    string title;
    int playtime;
    string content;
};

int getPlaytime(string begin, string end)
{
    int begin_minute = stoi(begin.substr(0,2))*60 + stoi(begin.substr(3,2));
    int end_minute = stoi(end.substr(0,2))*60 + stoi(end.substr(3,2));
    return end_minute - begin_minute;
}

string getContent(string partOfContent, int playtime)
{
    string content;
    string partOfContentSubstituted;
    for(int i=0;i<partOfContent.size();i++)
    {
        if(partOfContent[i]=='#')   partOfContentSubstituted.back()=tolower(partOfContentSubstituted.back());
        else    partOfContentSubstituted += partOfContent[i];
    }
    if(playtime==-1)    return  partOfContentSubstituted;
    for(int i=0;i<playtime/partOfContentSubstituted.size();i++)
    {
        content += partOfContentSubstituted;
    }
    content += partOfContentSubstituted.substr(0,playtime%partOfContentSubstituted.size());
    
    return content;
}

MusicInfo parse(string musicinfo)
{
    MusicInfo musicinfo_parsed;
    vector<int> indicesOfComma;
    int cur=-1;
    for(int i=0;i<3;i++)
    {
        cur = musicinfo.find_first_of(",",cur+1);
        indicesOfComma.push_back(cur);
    }
    musicinfo_parsed.playtime = getPlaytime(musicinfo.substr(0,indicesOfComma[0]),musicinfo.substr(indicesOfComma[0]+1,indicesOfComma[1]-indicesOfComma[0]-1));
    musicinfo_parsed.title = musicinfo.substr(indicesOfComma[1]+1,indicesOfComma[2]-indicesOfComma[1]-1);
    musicinfo_parsed.content = getContent(musicinfo.substr(indicesOfComma[2]+1), musicinfo_parsed.playtime);
    
    return musicinfo_parsed;
}

string solution(string m, vector<string> musicinfos) {
    string answer = "(None)";
    int playtimeOfAnswer=0;
    vector<MusicInfo>   musicinfos_parsed;
    for(int i=0;i<musicinfos.size();i++)
    {
        musicinfos_parsed.push_back(parse(musicinfos[i]));
    }
    for(int i=0;i<musicinfos_parsed.size();i++)
    {
        if(musicinfos_parsed[i].content.find(getContent(m,-1))!=-1 && musicinfos_parsed[i].playtime>playtimeOfAnswer) {
            answer = musicinfos_parsed[i].title;
            playtimeOfAnswer = musicinfos_parsed[i].playtime;
        }   
    }
    return answer;
}

File: Kakao/2018_방금그곡.swift
import Foundation

func substitue(chord: String)->String {
    var encoded = chord;
    let list = [("C#","c"),("D#","d"),("F#","f"),("G#","g"),("A#","a")]
    for i in 0..<list.count {
        encoded=encoded.replacingOccurrences(of: list[i].0, with: list[i].1)
    }
    return encoded
}

func getPlayTime(begin: String, end: String)->Int {
    let dateformatter = DateFormatter()
    dateformatter.dateFormat = "HH:mm"
    let beginTime = dateformatter.date(from: begin)!
    let endTime = dateformatter.date(from: end)!
    return Int(endTime.timeIntervalSince(beginTime))/60
}

func getOriginalChord(chord: String, playtime: Int)->String {
    let chord = chord.map({String($0)})
    var originalChord: String = ""
    for i in 0..<playtime {
        originalChord += chord[i%chord.count]
    }
    return originalChord
}

func solution(_ m:String, _ musicinfos:[String]) -> String {
    var answer = "(None)"
    var playTimeOfAnswer: Int = -1
    let chordListened = substitue(chord: m)
    
    for musicinfo in musicinfos {
        let musicinfo_parsed = musicinfo.split(separator: ",").map({String($0)})
        let playTime = getPlayTime(begin: musicinfo_parsed[0], end: musicinfo_parsed[1])
        let title = musicinfo_parsed[2]
        let chord = getOriginalChord(chord: substitue(chord: musicinfo_parsed[3]), playtime: playTime)
        if chord.contains(chordListened) && playTimeOfAnswer<playTime {
            answer = title
            playTimeOfAnswer = playTime
        }
    }
    return answer
}
